autogen/agentchat/conversable_agent.py

```diff
@@ -1330,3 +1333,77 @@ def can_execute_function(self, name: str) -> bool:
     def function_map(self) -> Dict[str, Callable]:
         """Return the function map."""
         return self._function_map
+
+    def function(
+        self,
+        *,
+        name: Optional[str] = None,
+        description: Optional[str] = None,
+        register_function: bool = True,
+    ) -> Callable[[F], F]:
+        """Decorator factory for registering a function to be used by an agent.
+
+        It's return value is used to decorate a function to be registered to the agent. The function uses type hints to
+        specify the arguments and return type. The function name is used as the default name for the function,
+        but a custom name can be provided. The function description is used to describe the function in the
+        agent's configuration.
+
+        Args:
+            name (optional(str)): name of the function. If None, the function name will be used (default: None).
+            description (optional(str)): description of the function (default: None). It is mandatory
+                for the initial decorator, but the following ones can omit it.
+            register_function (bool): whether to register the function to the agent (default: True)
+
+        Returns:
+            The decorator for registering a function to be used by an agent.
+
+        Examples:
+            ```
+            @agent2.function()
+            @agent1.function(description="This is a very useful function")
+            def my_function(a: Annotated[str, "description of a parameter"] = "a", b: int) -> str:
+                 return a + str(b)
+            ```
+
+        """
+
+        def _decorator(func: F) -> F:
+            """Decorator for registering a function to be used by an agent.
+
+            Args:
+                func: the function to be registered.
+
+            Returns:
+                The function to be registered, with the _description attribute set to the function description.
+
+            Raises:
+                ValueError: if the function description is not provided and not propagated by a previous decorator.
+
+            """
+            # name can be overwriten by the parameter, by default it is the same as function name
+            if name:
+                func._name = name
+            elif not hasattr(func, "_name"):
+                func._name = func.__name__
+
+            # description is propagated from the previous decorator, but it is mandatory for the first one
+            if description:
+                func._description = description
+            else:
+                if not hasattr(func, "_description"):
+                    raise ValueError("Function description is required, none found.")
+
+            # get JSON schema for the function
+            f = get_function_schema(func, name=func._name, description=func._description)
+
+            # register the function to the agent if there is LLM config, skip otherwise
+            if self.llm_config:
```

Comment:
if there is no llm config, should we throw instead of slient no op.

Reply:
Following the original example in `agentchat_function_call.ipynb`, user_proxy has no llm_config and it is used to execute functions, but not to propose calling them. In such cases, we need to call `self.register_function()`, but not to update `self.llm_config` as there is there is none.

```
user_proxy = autogen.UserProxyAgent(
    name="user_proxy",
    is_termination_msg=lambda x: x.get("content", "") and x.get("content", "").rstrip().endswith("TERMINATE"),
    human_input_mode="NEVER",
    max_consecutive_auto_reply=10,
    code_execution_config={"work_dir": "coding"},
)
```

Reply:
Correct. My mistake.